File: backend/src/schemas/jobs.py
"""
Pydantic schemas for extraction jobs system
"""
from datetime import datetime
from typing import Optional, Dict, Any, List
from uuid import UUID
from pydantic import BaseModel, Field, validator
from enum import Enum


class ScheduleType(str, Enum):
    """Job schedule types"""
    IMMEDIATE = "immediate"
    SCHEDULED = "scheduled"
    RECURRING = "recurring"


class JobStatus(str, Enum):
    """Job execution status"""
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"
    SKIPPED = "skipped"


class TriggeredBy(str, Enum):
    """How the job was triggered"""
    SCHEDULE = "schedule"
    MANUAL = "manual"
    IMMEDIATE = "immediate"


# ============================================================================
# BASE SCHEMAS
# ============================================================================

class JobScheduleConfig(BaseModel):
    """Schedule configuration for jobs"""
    cron: Optional[str] = Field(None, description="Cron expression for recurring jobs")
    timezone: Optional[str] = Field("UTC", description="Timezone for scheduling")
    run_at: Optional[datetime] = Field(None, description="Specific run time for scheduled jobs")
    
    @validator('cron')
    def validate_cron(cls, v):
        if v is not None:
            # Basic cron validation - could be enhanced with croniter
            parts = v.split()
            if len(parts) != 5:
                raise ValueError('Cron expression must have 5 parts')
        return v


class JobRetryPolicy(BaseModel):
    """Retry policy configuration"""
    max_retries: int = Field(3, ge=0, le=10, description="Maximum number of retries")
    retry_delay_minutes: int = Field(5, ge=1, le=60, description="Delay between retries in minutes")


# ============================================================================
# EXTRACTION JOB SCHEMAS
# ============================================================================

class ExtractionJobBase(BaseModel):
    """Base extraction job schema"""
    name: str = Field(..., min_length=1, max_length=255, description="Job name")
    description: Optional[str] = Field(None, description="Job description")
    category_id: UUID = Field(..., description="Document category ID")
    template_id: UUID = Field(..., description="Extraction template ID")
    schedule_type: ScheduleType = Field(..., description="Job schedule type")
    schedule_config: Optional[Dict[str, Any]] = Field(None, description="Schedule configuration")
    run_at: Optional[datetime] = Field(None, description="Specific run time for scheduled jobs")
    priority: int = Field(5, ge=1, le=10, description="Job priority (1=lowest, 10=highest)")
    max_concurrency: int = Field(5, ge=1, le=20, description="Max concurrent extractions")
    retry_policy: Optional[Dict[str, Any]] = Field(
        default={"max_retries": 3, "retry_delay_minutes": 5},
        description="Retry policy configuration"
    )
    is_active: bool = Field(True, description="Whether the job is active")


class ExtractionJobCreate(ExtractionJobBase):
    """Schema for creating extraction jobs"""
    
    @validator('schedule_config')
    def validate_schedule_config(cls, v, values):
        schedule_type = values.get('schedule_type')
        
        if schedule_type == ScheduleType.RECURRING:
            if not v or 'cron' not in v:
                raise ValueError('Recurring jobs must have cron expression in schedule_config')
        elif schedule_type == ScheduleType.SCHEDULED:
            if not values.get('run_at'):
                raise ValueError('Scheduled jobs must have run_at time specified')
        
        return v


class ExtractionJobUpdate(BaseModel):
    """Schema for updating extraction jobs"""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    description: Optional[str] = None
    schedule_type: Optional[ScheduleType] = None
    schedule_config: Optional[Dict[str, Any]] = None
    run_at: Optional[datetime] = None
    priority: Optional[int] = Field(None, ge=1, le=10)
    max_concurrency: Optional[int] = Field(None, ge=1, le=20)
    retry_policy: Optional[Dict[str, Any]] = None
    is_active: Optional[bool] = None


class ExtractionJobResponse(ExtractionJobBase):
    """Schema for extraction job responses"""
    id: UUID
    tenant_id: UUID
    last_run_at: Optional[datetime] = None
    next_run_at: Optional[datetime] = None
    total_executions: int = 0
    successful_executions: int = 0
    failed_executions: int = 0
    created_at: datetime
    updated_at: datetime
    
    # Related objects
    category_name: Optional[str] = None
    template_name: Optional[str] = None
    
    class Config:
        from_attributes = True


# ============================================================================
# DOCUMENT EXTRACTION TRACKING SCHEMAS
# ============================================================================

class DocumentExtractionTrackingBase(BaseModel):
    """Base document extraction tracking schema"""
    document_id: UUID
    job_id: UUID
    extraction_id: Optional[UUID] = None
    status: JobStatus
    triggered_by: TriggeredBy
    error_message: Optional[str] = None
    retry_count: int = 0


class DocumentExtractionTrackingCreate(DocumentExtractionTrackingBase):
    """Schema for creating document extraction tracking records"""
    pass


class DocumentExtractionTrackingUpdate(BaseModel):
    """Schema for updating document extraction tracking records"""
    status: Optional[JobStatus] = None
    extraction_id: Optional[UUID] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    processing_time_ms: Optional[int] = None
    error_message: Optional[str] = None
    retry_count: Optional[int] = None


class DocumentExtractionTrackingResponse(DocumentExtractionTrackingBase):
    """Schema for document extraction tracking responses"""
    id: UUID
    queued_at: datetime
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    processing_time_ms: Optional[int] = None
    created_at: datetime
    updated_at: datetime
    
    # Related objects
    document_filename: Optional[str] = None
    job_name: Optional[str] = None
    
    class Config:
        from_attributes = True


# ============================================================================
# LIST AND PAGINATION SCHEMAS
# ============================================================================

class ExtractionJobListResponse(BaseModel):
    """Schema for job list responses"""
    jobs: List[ExtractionJobResponse]
    total: int
    page: int
    per_page: int
    pages: int


class DocumentExtractionTrackingListResponse(BaseModel):
    """Schema for document tracking list responses"""
    tracking: List[DocumentExtractionTrackingResponse]
    total: int
    page: int
    per_page: int
    pages: int


# ============================================================================
# JOB EXECUTION SCHEMAS
# ============================================================================

class JobExecutionRequest(BaseModel):
    """Schema for manual job execution requests"""
    triggered_by: TriggeredBy = TriggeredBy.MANUAL
    priority_boost: Optional[int] = Field(None, ge=0, le=5, description="Temporary priority boost")


class JobExecutionResponse(BaseModel):
    """Schema for job execution responses"""
    job_id: UUID
    execution_started: bool
    documents_queued: int
    estimated_completion: Optional[datetime] = None
    message: str


# ============================================================================
# JOB STATISTICS SCHEMAS
# ============================================================================

class JobStatistics(BaseModel):
    """Job performance statistics"""
    total_executions: int
    successful_executions: int
    failed_executions: int
    success_rate: float
    average_processing_time_ms: Optional[int] = None
    last_execution: Optional[datetime] = None
    next_execution: Optional[datetime] = None


class JobAnalytics(BaseModel):
    """Job analytics and metrics"""
    job_id: UUID
    job_name: str
    statistics: JobStatistics
    recent_executions: List[DocumentExtractionTrackingResponse]
    documents_processed_today: int
    documents_pending: int


# ============================================================================
# ERROR SCHEMAS
# ============================================================================

class JobError(BaseModel):
    """Job error details"""
    job_id: UUID
    error_type: str
    error_message: str
    occurred_at: datetime
    retry_count: int
    max_retries: int
    next_retry_at: Optional[datetime] = None

File: frontend/src/services/api/jobs/types/jobs.ts
/**
 * Job Scheduling Types
 */
import { BaseEntity } from '../../base/types/common';

// Core Job Types
export interface Job extends BaseEntity {
  tenant_id: string;
  name: string;
  description?: string;
  job_type: 'extraction' | 'cleanup' | 'backup' | 'analytics' | 'custom';
  is_active: boolean;  // Backend uses is_active, not status
  schedule_type: 'immediate' | 'scheduled' | 'recurring' | 'manual';
  schedule_config: JobScheduleConfig;
  execution_config: JobExecutionConfig;
  retry_policy: JobRetryPolicy;
  last_run_at?: string;
  next_run_at?: string;
  created_by: string;
  // Statistics fields from backend
  total_executions?: number;
  successful_executions?: number;
  failed_executions?: number;
  // Additional fields from backend response
  category?: {
    id: string;
    name: string;
  };
  template?: {
    id: string;
    name: string;
    description?: string;
  };
  category_id?: string;
  template_id?: string;
}

export interface JobCreateRequest {
  name: string;
  description?: string;
  job_type: 'extraction' | 'cleanup' | 'backup' | 'analytics' | 'custom';
  schedule_type: 'immediate' | 'scheduled' | 'recurring' | 'manual';
  schedule_config: JobScheduleConfig;
  execution_config: JobExecutionConfig;
  retry_policy?: JobRetryPolicy;
}

export interface JobUpdateRequest {
  name?: string;
  description?: string;
  is_active?: boolean;  // Backend uses is_active, not status
  schedule_type?: 'immediate' | 'scheduled' | 'recurring' | 'manual';
  schedule_config?: JobScheduleConfig;
  execution_config?: JobExecutionConfig;
  retry_policy?: JobRetryPolicy;
}

export interface JobListParams {
  page?: number;
  per_page?: number;
  job_type?: string;
  is_active?: boolean;  // Backend uses is_active, not status
  schedule_type?: string;
  search?: string;
  sort_by?: string;
  sort_order?: 'asc' | 'desc';
  created_by?: string;
  category_id?: string;
  template_id?: string;
}

export interface JobListResponse {
  jobs: Job[];
  total: number;
  page: number;
  per_page: number;
  total_pages: number;
}

// Job Schedule Configuration
export interface JobScheduleConfig {
  cron?: string;  // Backend uses 'cron', not 'cron_expression'
  cron_expression?: string;  // Keep for backward compatibility
  timezone?: string;
  run_at?: string;
  interval_minutes?: number;
  days_of_week?: number[];
  days_of_month?: number[];
  start_date?: string;
  end_date?: string;
  max_runs?: number;
  max_runtime_hours?: number;
}

// Job Execution Configuration
export interface JobExecutionConfig {
  template_id?: string;
  category_id?: string;
  document_filters?: Record<string, any>;
  extraction_settings?: Record<string, any>;
  cleanup_settings?: {
    older_than_days: number;
    keep_successful: boolean;
    keep_failed: boolean;
  };
  backup_settings?: {
    include_documents: boolean;
    include_extractions: boolean;
    include_templates: boolean;
    compression: boolean;
  };
  analytics_settings?: {
    metrics: string[];
    date_range: {
      start: string;
      end: string;
    };
  };
  custom_settings?: Record<string, any>;
}

// Job Retry Policy
export interface JobRetryPolicy {
  max_retries: number;
  retry_delay_minutes: number;
  backoff_multiplier: number;
  max_retry_delay_minutes: number;
  retry_on_failure_types: string[];
}

// Job Execution
export interface JobExecution extends BaseEntity {
  job_id: string;
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled' | 'timeout';
  started_at?: string;
  completed_at?: string;
  duration_ms?: number;
  progress_percentage: number;
  result_data?: Record<string, any>;
  error_message?: string;
  retry_count: number;
  triggered_by: 'schedule' | 'manual' | 'api';
  triggered_by_user?: string;
}

export interface JobsExecutionRequest {
  triggered_by: 'manual' | 'api';
  parameters?: Record<string, any>;
  priority?: 'low' | 'normal' | 'high' | 'urgent';
}

export interface JobsExecutionResponse {
  execution_id: string;
  job_id: string;
  status: string;
  message: string;
  estimated_duration_ms?: number;
  queue_position?: number;
}

// Job History and Statistics
export interface JobHistoryResponse {
  executions: JobExecution[];
  total: number;
  page: number;
  per_page: number;
  total_pages: number;
}

export interface JobStatistics {
  total_executions: number;
  successful_executions: number;
  failed_executions: number;
  cancelled_executions: number;
  timeout_executions: number;
  success_rate: number;
  avg_execution_time_ms: number;
  total_execution_time_ms: number;
  last_execution_at?: string;
  next_execution_at?: string;
  execution_trend: Array<{
    date: string;
    executions: number;
    success_rate: number;
    avg_duration_ms: number;
  }>;
}

// Job Monitoring
export interface JobMonitor {
  job_id: string;
  status: 'running' | 'paused' | 'stopped';
  current_execution?: JobExecution;
  queue_position: number;
  estimated_start_time?: string;
  resource_usage: {
    cpu_percent: number;
    memory_mb: number;
    disk_io_mb: number;
  };
  progress: {
    completed_steps: number;
    total_steps: number;
    current_step: string;
    estimated_completion?: string;
  };
}

export interface JobQueueStatus {
  total_jobs: number;
  running_jobs: number;
  queued_jobs: number;
  failed_jobs: number;
  avg_queue_time_ms: number;
  estimated_completion_time?: string;
  system_load: {
    cpu_percent: number;
    memory_percent: number;
    disk_percent: number;
  };
}

// Job Templates
export interface JobTemplate {
  id: string;
  name: string;
  description: string;
  job_type: string;
  default_schedule_config: JobScheduleConfig;
  default_execution_config: JobExecutionConfig;
  default_retry_policy: JobRetryPolicy;
  is_system_template: boolean;
  created_by: string;
  created_at: string;
  updated_at: string;
}

export interface JobTemplateCreateRequest {
  name: string;
  description: string;
  job_type: string;
  default_schedule_config: JobScheduleConfig;
  default_execution_config: JobExecutionConfig;
  default_retry_policy: JobRetryPolicy;
}

// Job Validation
export interface JobValidationResult {
  is_valid: boolean;
  errors: string[];
  warnings: string[];
  suggestions: string[];
}

// Job Status Types
export type JobStatus = 'active' | 'inactive' | 'paused' | 'completed' | 'failed';

export type JobType = 'extraction' | 'cleanup' | 'backup' | 'analytics' | 'custom';

export type ScheduleType = 'immediate' | 'scheduled' | 'recurring' | 'manual';

export type ExecutionStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled' | 'timeout';

export type TriggerType = 'schedule' | 'manual' | 'api';

export type JobPriority = 'low' | 'normal' | 'high' | 'urgent';

// Job Constants
export const DEFAULT_RETRY_POLICY: JobRetryPolicy = {
  max_retries: 3,
  retry_delay_minutes: 5,
  backoff_multiplier: 2,
  max_retry_delay_minutes: 60,
  retry_on_failure_types: ['timeout', 'network_error', 'temporary_failure'],
};

export const JOB_TYPES: JobType[] = ['extraction', 'cleanup', 'backup', 'analytics', 'custom'];

export const SCHEDULE_TYPES: ScheduleType[] = ['immediate', 'scheduled', 'recurring', 'manual'];

export const EXECUTION_STATUSES: ExecutionStatus[] = ['pending', 'running', 'completed', 'failed', 'cancelled', 'timeout'];

export const JOB_PRIORITIES: JobPriority[] = ['low', 'normal', 'high', 'urgent'];

// Cron Expression Examples
export const CRON_EXAMPLES = {
  'Every minute': '*/1 * * * *',
  'Every 5 minutes': '*/5 * * * *',
  'Every hour': '0 * * * *',
  'Every day at midnight': '0 0 * * *',
  'Every weekday at 9 AM': '0 9 * * 1-5',
  'Every Sunday at 2 AM': '0 2 * * 0',
  'Every month on the 1st': '0 0 1 * *',
  'Every quarter': '0 0 1 1,4,7,10 *',
  'Every year on January 1st': '0 0 1 1 *',
};
